util/jemalloc_nodump_allocator.cc

```diff
@@ -48,12 +120,41 @@ void* JemallocNodumpAllocator::Alloc(extent_hooks_t* extent, void* new_addr,
   return result;
 }
 
-JemallocNodumpAllocator::~JemallocNodumpAllocator() {
-  assert(arena_index_ != 0);
-  std::string key = "arena." + ToString(arena_index_) + ".destroy";
+Status JemallocNodumpAllocator::DestroyArena(unsigned arena_index) {
+  assert(arena_index != 0);
+  std::string key = "arena." + ToString(arena_index) + ".destroy";
   int ret = mallctl(key.c_str(), nullptr, 0, nullptr, 0);
   if (ret != 0) {
-    fprintf(stderr, "Failed to destroy jemalloc arena, error code: %d\n", ret);
+    return Status::Incomplete("Failed to destroy jemalloc arena, error code: " +
+                              ToString(ret));
+  }
+  return Status::OK();
+}
+
+void JemallocNodumpAllocator::DestroyThreadSpecificCache(void* ptr) {
+  assert(ptr != nullptr);
+  unsigned* tcache_index = static_cast<unsigned*>(ptr);
+  size_t tcache_index_size = sizeof(unsigned);
+  int ret __attribute__((__unused__)) =
+      mallctl("tcache.destroy", tcache_index, &tcache_index_size, nullptr, 0);
+  // Silently ignore error.
+  assert(ret == 0);
+  delete tcache_index;
+}
+
+JemallocNodumpAllocator::~JemallocNodumpAllocator() {
+  // Destroy tcache before destroying arena.
```

Comment:
@interwq Is it required to destroy tcache before destroying arenas?

Reply:
Yes if the tcache may hold items from the arena (i.e. if you combined the tcache id with the specific arena.

Note that, there's also tcache flush (http://jemalloc.net/jemalloc.3.html#tcache.flush), which releases all items in the cache but keeps the tcache id active, in case you want to keep it around.